assembly/src/assembler/mod.rs

```diff
@@ -595,86 +599,154 @@ impl Assembler {
+                        BasicBlockOrDecorators::Nothing => None,
+                    };
 
-                    let loop_body_node_id =
-                        self.compile_body(body.iter(), proc_ctx, None, mast_forest_builder)?;
+                    let loop_node_id = {
+                        let loop_body_node_id =
+                            self.compile_body(body.iter(), proc_ctx, None, mast_forest_builder)?;
+                        mast_forest_builder.ensure_loop(loop_body_node_id)?
+                    };
+                    if let Some(pre_decorator_ids) = maybe_pre_decorators {
+                        mast_forest_builder.set_before_enter(loop_node_id, pre_decorator_ids)
+                    }
 
-                    let loop_node_id = mast_forest_builder.ensure_loop(loop_body_node_id)?;
-                    node_ids.push(loop_node_id);
+                    body_node_ids.push(loop_node_id);
                 },
             }
         }
 
-        if let Some(basic_block_id) =
-            basic_block_builder.try_into_basic_block(mast_forest_builder)?
-        {
-            node_ids.push(basic_block_id);
-        }
+        let maybe_post_decorators: Option<Vec<DecoratorId>> =
+            match basic_block_builder.try_into_basic_block(mast_forest_builder)? {
+                BasicBlockOrDecorators::BasicBlock(basic_block_id) => {
+                    body_node_ids.push(basic_block_id);
+                    None
+                },
+                BasicBlockOrDecorators::Decorators(decorator_ids) => {
+                    // the procedure body ends with a list of decorators
+                    Some(decorator_ids)
+                },
+                BasicBlockOrDecorators::Nothing => None,
+            };
+
+        let procedure_body_id = if body_node_ids.is_empty() {
+            // We cannot allow only decorators in a procedure body, since decorators don't change
+            // the MAST digest of a node. Hence, two empty procedures with different decorators
+            // would look the same to the `MastForestBuilder`.
+            if maybe_post_decorators.is_some() {
+                return Err(AssemblyError::EmptyProcedureBodyWithDecorators {
+                    span: proc_ctx.span(),
+                    source_file: proc_ctx.source_manager().get(proc_ctx.span().source_id()).ok(),
+                })?;
+            }
 
-        Ok(if node_ids.is_empty() {
             mast_forest_builder.ensure_block(vec![Operation::Noop], None)?
         } else {
-            mast_forest_builder.join_nodes(node_ids)?
-        })
+            mast_forest_builder.join_nodes(body_node_ids)?
+        };
+
+        // Make sure that any post decorators are added at the end of the procedure body
+        if let Some(post_decorator_ids) = maybe_post_decorators {
+            mast_forest_builder.set_after_exit(procedure_body_id, post_decorator_ids);
+        }
```

Comment:
If I understood it correctly, the only time we are setting the `after_exit` decorators is at the end of procedure body. I wonder if there is some clever way to get rid of it `after_exit` decorators altogether and append these somewhere else. Not sure if this is possible though.

Reply:
As discussed before implementing #1482, I don't think we can get rid of `after_exit`